client/src/map/trackingContainer.tsx

```diff
@@ -0,0 +1,89 @@
+import React, { useEffect, useState } from 'react';
+
+import Button from '../common/button';
+import Modal, {
+  ModalParagraph,
+  ModalButtonRow,
+  ModalHeader,
+} from '../common/modal';
+import fetchName from '../location/webbluetooth';
+import BluetoothNameModal from './bluetoothNameModal';
+import { BeaconGeoLocation } from '../location/mqttDeserialize';
+
+interface Props {
+  confirmName(a: string): void;
+  beacons: BeaconGeoLocation[];
+  onStaticSelected(a: string): void;
+  onClose(): void;
+}
+
+const TrackingContainer = ({ onClose, confirmName, beacons }: Props) => {
+  const [bluetoothLoading, setBluetoothLoading] = useState(false);
+  const [manualDeviceSelect, setManualDeviceSelect] = useState(false);
+
+  const [isOpen, setIsOpen] = useState(true);
+  const [nameSelection, setNameSelection] = useState<null | string>(null);
+
+  useEffect(() => {
+    return () => {
+      setIsOpen(false);
+    };
+  }, []);
+
+  if (manualDeviceSelect) {
+    return (
+      <BluetoothNameModal
+        isOpen={isOpen}
+        closeModal={onClose}
+        setNameSelection={setNameSelection}
+        beacons={beacons}
+        nameSelection={nameSelection}
+        confirmName={confirmName}
+      />
+    );
+  }
+
+  return (
+    <Modal isOpen={isOpen} onRequestClose={onClose}>
+      <ModalHeader>Ubikampus indoor positioning</ModalHeader>
+      {bluetoothLoading && <span>loading...</span>}
+      <ModalParagraph>Allow Ubimaps to track my location</ModalParagraph>
+      <ModalButtonRow>
+        <Button
+          onClick={async () => {
+            setBluetoothLoading(true);
+            try {
+              const name = await fetchName();
+
+              console.log('got bt name', name);
+              setBluetoothLoading(false);
+              confirmName(name);
+            } catch (err) {
+              if (err.message && err.message.includes('User cancelled')) {
+                onClose();
```

Comment:
I think we could still consider what happens, if the user cancels the device pairing operation.

Would it be sensible, then, to default to manual device selection? Or, should we, rather, offer the option of selecting the name manually, at the user's discretion. In other words, would it make sense to have the options "automatic" and "manual".

I'm thinking this from my current perspective. I'm currently working outside the Ubikampus space, without access to the Raspberry Pis. Nevertheless, the system defaults to automatic device selection. In other words, the system tries to form a pair with a Raspberry Pi. Manual device selection is not available as an option although, in this case, that would be nice. Of course I can look up the code and change the default... So from the point of view of the coder this is a minor issue.

I wonder if the end user has any credible use cases as well? For instance, if somebody feels uneasy about pairing their device with our Raspberry Pis.

Just some food for thought. I'll leave this to your discretion.

Reply:
Good points, definitely things i didn't think of before.

One detail i was hoping would be possible was that the application would fully hide the device selection (manual option) functionality for users where the device name fetch succeeds. The idea was that the user flow would be very smooth for users where the web bluetooth works, and these users would never know about the dirty device selection workaround.

Additionally a dialog between manual and automatic device selection would require some instructions for the user, they might get confused why the app wants to know the device bluetooth name.

For testing the user flow with manual device selection, i think best way is to just use browser without web bluetooth functionality, like Firefox.